Para la practica de entornos de desarrollo usaremos Vagrant como administrador de ambientes y a Ansible como provisionador.

# Boxes

Boxes son paquetes que contienen un sistema o estructura base para los ambientes de Vagrant.

## Administración manual de boxes

Vagrant provee el comando `box` para la adminstración de las boxes.

In [4]:
vagrant box --help

Usage: vagrant box <subcommand> [<args>]

Available subcommands:
     add
     list
     outdated
     remove
     repackage
     update

For help on any individual subcommand run `vagrant box <subcommand> -h`



## Agregar la caja de Ubuntu

Debemo de proveer el identificador, la url o la ruta de la caja.

In [5]:
vagrant box add ubuntu/trusty64

==> box: Loading metadata for box 'ubuntu/trusty64'
    box: URL: https://atlas.hashicorp.com/ubuntu/trusty64
==> box: Adding box 'ubuntu/trusty64' (v20160708.1.2) for provider: virtualbox
    box: Downloading: https://atlas.hashicorp.com/ubuntu/boxes/trusty64/versions/20160708.1.2/providers/virtualbox.box
==> box: Successfully added box 'ubuntu/trusty64' (v20160708.1.2) for 'virtualbox'!


## Lista de boxes

In [6]:
vagrant box list

ubuntu/trusty64 (virtualbox, 20160708.1.2)


# Providers

Los provedores son los que se encargan de crear, pausar, elminar los ambientes de desarollo.

Vagrant viene con soporte para VMware, VirtualBox y Docker pero otros provedores se pueden integrar.

# Vagrantfile

Es un archivo que indica la maquina requerida por el proyecto y como configurar y provisionar la maquina.

La sintaxis usada es la misma que en ruby, el lenguaje de programación pero no es necesario tener
conocimiento previo de este, ya que solo se trata de asignación de variables.

In [10]:
cat Vagrantfile

Vagrant.configure(2) do |config|
  config.vm.box = "ubuntu/trusty64"
end


## Creación de ambiente

Para crear un ambiente solo necesitas ejecutar el comando up de vagrant.

In [11]:
vagrant up

Bringing machine 'default' up with 'virtualbox' provider...
==> default: Importing base box 'ubuntu/trusty64'...
==> default: Matching MAC address for NAT networking...
==> default: Checking if box 'ubuntu/trusty64' is up to date...
==> default: Setting the name of the VM: hacker_school_default_1469015343896_77366
==> default: Clearing any previously set forwarded ports...
==> default: Clearing any previously set network interfaces...
==> default: Preparing network interfaces based on configuration...
    default: Adapter 1: nat
==> default: Forwarding ports...
    default: 22 (guest) => 2222 (host) (adapter 1)
==> default: Booting VM...
==> default: Waiting for machine to boot. This may take a few minutes...
    default: SSH address: 127.0.0.1:2222
    default: SSH username: vagrant
    default: SSH auth method: private key
    default: 
    default: Vagrant insecure key detected. Vagrant will automatically replace
    default: this with a newly generated keypair for better security.


## Interactuar con el ambiente

Ejecuta en tu consola el comando ssh para establecer comunicación a travez del puerto 2222 del host usando una llave insegura, estos elementos se pueden personalizar en el Vagrantfile.

In [12]:
vagrant ssh -c 'uname -a'

Linux vagrant-ubuntu-trusty-64 3.13.0-92-generic #139-Ubuntu SMP Tue Jun 28 20:42:26 UTC 2016 x86_64 x86_64 x86_64 GNU/Linux



Para una lista de cajas listas para usar pueder acceder a http://www.vagrantbox.es/, agrega el nombre y la url de la maquina con el comando `add` antes mencionado.

# Provisionado

Se sirve de herramientas que permiten instalar y modificar los ambientes base y ajustarlos a nuestras necesidades.

Vagrant trae soporte para varios sistemas como son Ansible, Pupet, Salt, Bash o varios sabores de Chef.
De momento trabajaremos con ansible ya que es muy sencillo de aprender.

## Ansible

Ansible mantiene una estructura muy simple que puede extenderse a gusto, sirviendo desde un pequeño nodo hasta miles. La unidad base de declaraciones son los playbook, arcivos donde se declara la condiguración deseada para el sistema. LA sintaxis es muy simple y no es más que un archivo con YAML.

Agregamos una sección donde especificamos el provisionador que queremos usar, ademas le indicamos que playbook va a usar para el provisionamiento del sistema.

In [22]:
cat Vagrantfile

Vagrant.configure(2) do |config|
  config.vm.box = "ubuntu/trusty64"

  config.vm.provision "ansible" do |ansible|
    ansible.playbook = "playbook.yml"
  end
end


In [20]:
vagrant provision

==> default: Running provisioner: ansible...
    default: Running ansible-playbook...

PLAY [all] *********************************************************************

TASK [setup] *******************************************************************
ok: [default]

TASK [Instalación de python] ***************************************************
ok: [default]

PLAY RECAP *********************************************************************
default                    : ok=2    changed=0    unreachable=0    failed=0   



In [21]:
cat playbook.yml

---
- hosts: all
  tasks:
      - name: Instalación de python
        apt: pkg=python state=installed update_cache=true
        become: True



## Practica

1. Crea un nuevo Vagranfile
2. Escoge y agrega el caja del sistema de tu agrado (Ubuntu, Fedora, Windows)
3. Provisiona la maquina con python, mongo, git
4. Inicia el servidor manualmente y que muestre un hola mundo usando `python -m SimpleHTTPServer 80` 
5. La aplicación debe mostrarse cuando en el navegador local use 127.0.0.1:8080
6. Punto extra si agregas livereload o creas un servicio para tu servidor

## Detener y eliminar una maquina

In [ ]:
vagrant halt

In [ ]:
vagrant destroy

Recursos:

- [Charla de Ansible](http://zoek1.github.io/ninja-code-party-2014-ansible-101)
- [Sitio de Vagrant](https://www.vagrantup.com/docs/plugins/)
- [Listado de boxes](http://www.vagrantbox.es/)
